### PR TITLE
Add e2e case for TiDB Operator upgrade and check database

### LINKED ISSUE #4441: Add e2e case for TiDB Operator upgrade
## Feature Request

**Is your feature request related to a problem? Please describe:**

Add e2e case for TiDB Operator upgrade, refer to https://github.com/pingcap/tidb-operator/pull/4430.
- Deploy TiDB Operator v1.2.7 + TiDB/TiFlash v5.3.0
  - Check TiDB cluster is ready
  - Prepare 4 tables with some data
  - Enable TiFlash replica for table 1, check that replica is ready in TiFlash and data in table 1 can be queried from TiFlash
  - Data in table 2 can be queried from TiKV, insert some data to table 2
  - Upgrade to the TiDB Operator with this fix, no rolling update occurs
  - Enable TiFlash replica for table 3, check that replica is ready in TiFlash and data in table 1,3 can be queried from TiFlash
  - Data in table 2 can be queried from TiKV, insert some data to table 2
  - Upgrade TiDB to the latest version in e2e, e.g. v5.4.0, TiDB cluster is ready, enable TiFlash replica for table 4, check that replica is ready in TiFlash and data in table 1,3,4 can be queried from TiFlash.
  - Data in table 2 can be queried from TiKV, insert some data to table 2
**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by KanShiori**
### What problem does this PR solve?

Close #4441

### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [x] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```

### CODE REVIEW

File: tests/e2e/util/db/db.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+func (d *Database) TiFlashAction() *TiFlashAction {
+	return &TiFlashAction{
+		baseAction: &baseAction{
+			db: d,
+		},
+	}
+}
+
+func (d *Database) Ping() error {
+	return d.driver.Ping()
```
Comment (csuzhangxc) [inline tests/e2e/util/db/db.go:93]: How about calling `Ping` in `NewDatabase`?

File: tests/e2e/tidbcluster/serial.go [hunk 2]
```diff
@@ -690,6 +694,95 @@ var _ = ginkgo.Describe("[Serial]", func() {
 			})
 		})
 
+		ginkgo.Context("From prev major version", func() {
+
+			ginkgo.BeforeEach(func() {
+				operatorVersion = OperatorPrevMajorVersion
+			})
+
+			ginkgo.It("should not trigger rolling-update and database work fine", func() {
```
Comment (csuzhangxc) [inline tests/e2e/tidbcluster/serial.go:703]: `We need to cover the TLS case.`, https://github.com/pingcap/tidb-operator/issues/4441#issuecomment-1053807798

File: tests/e2e/util/db/db.go (hunk 1 above)
Comment (KanShiori) [reply tests/e2e/util/db/db.go:93]: Do we need to ping DB when open？

Comment (csuzhangxc) [reply tests/e2e/util/db/db.go:93]: https://github.com/golang/go/blob/29b968e76638c22368f775a4347a544a096d9380/src/database/sql/sql.go#L808-L810

Comment (KanShiori) [reply tests/e2e/util/db/db.go:93]: Get 👀

File: tests/e2e/tidbcluster/serial.go (hunk 2 above)
Comment (KanShiori) [reply tests/e2e/tidbcluster/serial.go]: Done

**Review by csuzhangxc — APPROVED** [review]

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**